### PR TITLE
feat: add --rpc-hostname option to datanode for a persist address to store in meta

### PR DESCRIPTION
**Pull request by sunng87**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

At the moment, we stored datanode's grpc IP address in meta. However when this address changes, via a deployment in k8s for example, it's impossible for frontend to find that datanode again. This patch adds a new start option `--rpc-hostname` to `datanode` startup command and it will use value to report to meta.

- this option is optional and it fallback to `--rpc-addr`, this behaviour is completely same to current
- as its name suggests, you can configure hostname and it resolves port from `--rpc-addr`
- user has to guarantee that the `--rpc-hostname` resolves to `--rpc-addr` so that frontend can call datanode via this hostname. User can use `0.0.0.0` in `--rpc-addr` as an exception, if the environment permits

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by sunng87** [comment]
The option name may be confusing because it requires port as well. Or we can try to resolve port from `--rpc-addr` for better usability.

File: src/datanode/src/heartbeat.rs [hunk 1]
```diff
@@ -142,3 +145,44 @@ impl HeartbeatTask {
...
+fn resolve_addr(bind_addr: &str, hostname_addr: &Option<String>) -> String {
+    match hostname_addr {
+        Some(hostname_addr) => {
+            // it has port configured
+            if hostname_addr.contains(':') {
+                hostname_addr.clone()
+            } else {
+                // otherwise, resolve port from bind_addr
+                // should be safe to unwrap here because bind_addr is already validated
+                let port = &bind_addr.split(':').nth(1).unwrap();
```
Comment (Fengys123) [inline src/datanode/src/heartbeat.rs:160]:
```suggestion
                let port = bind_addr.split(':').nth(1).unwrap();
```

File: src/datanode/src/heartbeat.rs [hunk 2]
```diff
@@ -142,3 +145,44 @@ impl HeartbeatTask {
...
+                format!("{hostname_addr}:{port}")
+            }
+        }
+        None => bind_addr.to_owned(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
```
Comment (Fengys123) [inline src/datanode/src/heartbeat.rs:170]: nit:
```suggestion
```

**Review by Fengys123 — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**